File: SecondYear/ApoorvaSri123_ApoorvaSrivastava_2226it1116_2/Week_4/minimum_parenthesis.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int minAddToMakeValid(char* pattern) {
    int openCount = 0;
    int addCount = 0;

    for (int i = 0; pattern[i] != '\0'; i++) {
        if (pattern[i] == '(') {
            openCount++;
        } else if (pattern[i] == ')') {
            if (openCount > 0) {
                openCount--;
            } else {
                addCount++;
            }
        }
    }

    return openCount + addCount;
}

int main() {
    int numPatterns;
    scanf("%d", &numPatterns);
    getchar(); // Consume the newline character

    int* results = (int*)malloc(numPatterns * sizeof(int));

    for (int i = 0; i < numPatterns; i++) {
        char pattern[1000]; // Adjust the size as needed
        fgets(pattern, sizeof(pattern), stdin);

        // Remove the newline character from the input
        pattern[strcspn(pattern, "\n")] = '\0';

        int minAdditions = minAddToMakeValid(pattern);
        results[i] = minAdditions;
    }

    for (int i = 0; i < numPatterns; i++) {
        printf("%d\n", results[i]);
    }

    free(results);
    return 0;
}

File: SecondYear/praveer7398_PraveerNandan_2226ce1011_2/Week_4/minimum_parenthesis.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;

public class minimum_parenthesis {

    public static int minAddToMakeValid(String pattern) {
        int openCount = 0;
        int addCount = 0;

        for (char c : pattern.toCharArray()) {
            if (c == '(') {
                openCount++;
            } else if (c == ')') {
                if (openCount > 0) {
                    openCount--;
                } else {
                    addCount++;
                }
            }
        }

        return openCount + addCount;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        
        int numPatterns = scanner.nextInt();
        scanner.nextLine(); 

        List<Integer> results = new ArrayList<>();

        for (int i = 1; i <= numPatterns; i++) {
            
            String pattern = scanner.nextLine();

            int minAdditions = minAddToMakeValid(pattern);
            results.add(minAdditions);
        }

        
        for (int i = 0; i < numPatterns; i++) {
            System.out.println(results.get(i));
        }

        scanner.close();
    }
}
